### PR TITLE
cmd/atlas: added context flag to migrate apply

### PR DESCRIPTION
**Pull request by dorav**

### CODE REVIEW

File: cmd/atlas/internal/cmdapi/migrate.go [hunk 1]
```diff
@@ -111,6 +113,11 @@ If run with the "--dry-run" flag, atlas will not execute any SQL.`,
 							set.Flush(cmd, cmdErr)
 						}
 					}()
+					if flags.context != "" {
+						if err := json.Unmarshal([]byte(flags.context), &set.Context); err != nil {
+							return fmt.Errorf("invalid --context: %w", err)
+						}
+					}
```
Comment (a8m) [inline cmd/atlas/internal/cmdapi/migrate.go:120]: Handle it in the set creation (you can pass the flags to it).

File: cmd/atlas/internal/cmdapi/migrate.go [hunk 2]
```diff
@@ -134,6 +141,7 @@ If run with the "--dry-run" flag, atlas will not execute any SQL.`,
 	cmd.Flags().StringVarP(&flags.txMode, flagTxMode, "", txModeFile, "set transaction mode [none, file, all]")
 	cmd.Flags().BoolVarP(&flags.allowDirty, flagAllowDirty, "", false, "allow start working on a non-clean database")
 	cmd.MarkFlagsMutuallyExclusive(flagLog, flagFormat)
+	migrateApplySetFlags(cmd, &flags)
```
Comment (a8m) [inline cmd/atlas/internal/cmdapi/migrate.go:144]: I think we can simply inline the logic here. Most of the code already resides in this file.

File: cmd/atlas/internal/cmdapi/migrate.go [hunk 3]
```diff
@@ -273,6 +281,7 @@ type (
 		log    *cmdlog.MigrateApply
 		rrw    cmdmigrate.RevisionReadWriter
 		done   func(*cloudapi.ReportMigrationInput)
+		set    *MigrateReportSet
```
Comment (a8m) [inline cmd/atlas/internal/cmdapi/migrate.go:284]: Why it's necessary?

**Review by a8m — COMMENTED** [review]
See comments and add tests for this please

File: cmd/atlas/internal/cmdapi/migrate.go (hunk 2 above)
Comment (dorav) [reply cmd/atlas/internal/cmdapi/migrate.go:144]: Love this ❤️

File: cmd/atlas/internal/cmdapi/migrate.go [hunk 4]
```diff
@@ -435,9 +447,10 @@ func (r *MigrateReport) Done(cmd *cobra.Command, flags migrateApplyFlags) error
 		return logApply(cmd, cmd.OutOrStdout(), flags, r.log)
 	}
 	var (
-		ver  string
-		clog bytes.Buffer
-		err  = logApply(cmd, io.MultiWriter(cmd.OutOrStdout(), &clog), flags, r.log)
+		ver        string
+		clog       bytes.Buffer
+		err        = logApply(cmd, io.MultiWriter(cmd.OutOrStdout(), &clog), flags, r.log)
+		setContext *cloudapi.DeployContextInput
```
Comment (a8m) [inline cmd/atlas/internal/cmdapi/migrate.go:453]: Why it's here?

File: cmd/atlas/internal/cmdapi/migrate_test.go [hunk 5]
```diff
@@ -877,13 +878,14 @@ env {
...
 		_, err = uuid.Parse(report.Target.ID)
-		require.NoError(t, err, "target id is not a valid uuid")
-		require.False(t, report.StartTime.IsZero())
-		require.False(t, report.EndTime.IsZero())
-		require.Equal(t, cloudapi.ReportMigrationInput{
+		assert.NoError(t, err, "target id is not a valid uuid")
+		assert.False(t, report.StartTime.IsZero())
+		assert.False(t, report.EndTime.IsZero())
+		assert.Nil(t, report.Context)
+		assert.Equal(t, cloudapi.ReportMigrationInput{
```
Comment (a8m) [inline cmd/atlas/internal/cmdapi/migrate_test.go:888]: Don't change existing tests, please

File: cmd/atlas/internal/cmdapi/migrate_test.go [hunk 6]
```diff
@@ -11,6 +11,7 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"github.com/stretchr/testify/assert"
```
Comment (a8m) [inline cmd/atlas/internal/cmdapi/migrate_test.go:14]:
```suggestion
```
Use "require" instead. I prefer the "fail fast" approach, and that's why "require" is the main used here.

File: cmd/atlas/internal/cmdapi/migrate_test.go [hunk 7]
```diff
@@ -900,6 +902,27 @@ env {
...
+		cmd.AddCommand(migrateApplyCmd())
+		s, err := runCmd(
+			cmd, "apply",
+			"-c", "file://"+path,
+			"--env", "local",
+			"--url", u,
+			"--var", "cloud_url="+srv.URL,
+			"--context", `{ "triggerType": "KUBERNETES", "triggerVersion": "v1.2.3" }`,
+			// Inject fake variable to enforce re-evaluation of the data source (skip cache).
+			"--var", fmt.Sprintf("cache=%s", uuid.NewString()),
```
Comment (a8m) [inline cmd/atlas/internal/cmdapi/migrate_test.go:916]: not sure you need this here (you're not checking the data source)

File: cmd/atlas/internal/cmdapi/migrate_test.go [hunk 8]
```diff
@@ -1124,6 +1152,26 @@ env {
...
+			"--var", fmt.Sprintf("urls=%s", openSQLite(t, "")),
+			"--var", fmt.Sprintf("urls=%s", openSQLite(t, "")),
+			"--var", "cloud_url="+srv.URL,
+		)
+		require.NoError(t, err)
+		require.Equal(t, "No migration files to execute\nNo migration files to execute\nhttps://gh.atlasgo.cloud/deployments/sets/94489280524\n", s)
+		if assert.NotNil(t, report.Context) {
+			assert.Equal(t, report.Context.TriggerType, "KUBERNETES")
+			assert.Equal(t, report.Context.TriggerVersion, "v1.2.3")
+		}
```
Comment (a8m) [inline cmd/atlas/internal/cmdapi/migrate_test.go:1172]: ditto

**Review by a8m — COMMENTED** [review]
See comments. Also, please let's not change existing tests as there's no need for this in this PR.

**Review by a8m — APPROVED** [review]

**Pull request merged**